### ISSUE TITLE
Realm is still not mockable with Mockito after 2.2.2 change

### ISSUE THREAD

**Issue by NeverwinterMoon** [realm/realm-java#3937]
So, as discussed here: https://github.com/realm/realm-java/issues/3869 and implemented since 2.2.2 (https://github.com/realm/realm-java/pull/3911/files), the removal of the final modifier would allow mocking Realm without the use of PowerMock. Needless to say, after spending so much time trying to make Realm + PowerMock + Roboelectric work (and failing to do so), I was ecstatic to hear about aforementioned changes. But...

It was still impossible to mock Realm.class:
`Mockito.mock(Realm.class)` would end up
```
org.mockito.exceptions.base.MockitoException: 
Mockito cannot mock this class: class io.realm.Realm.

Mockito can only mock non-private & non-final classes.
If you're not sure why you're getting this error, please report to the mailing list.


Java               : 1.8
JVM vendor name    : Oracle Corporation
JVM vendor version : 25.112-b15
JVM name           : Java HotSpot(TM) 64-Bit Server VM
JVM version        : 1.8.0_112-b15
JVM info           : mixed mode
OS name            : Windows 7
OS version         : 6.1
```

My guess is it's because Realm.class still has a private constructor.

Am I doing something wrong and there is a more elaborate way to set things up? Is there still a way to mock stuff without PowerMock? If not, what is the point of https://github.com/realm/realm-java/pull/3911/files change?

Things I used in my code: 
- `io.realm:realm-gradle-plugin:2.3.0-SNAPSHOT`
- `org.mockito:mockito-core:2.4.0`
- `com.google.dagger:dagger:2.8`
- `org.robolectric:robolectric:3.1.4`

Oh, and all the production and test code is written in Kotlin.

**Comment by cmelchior** [comment]
Have you made sure you are using the latest SNAPSHOT by using `./gradlew clean assemble --refresh-dependencies` ?

**Comment by NeverwinterMoon** [comment]
@cmelchior Yes. In fact, I completely cleaned everything, including the Gradle cache during one of the attempts. Nonetheless, just to be sure, I did `./gradlew clean assemble --refresh-dependencies` just now. Still the same result.

**Comment by NeverwinterMoon** [comment]
@cmelchior OK, I see the problem. It doesn't work with Mockito 2.x and works with 1.x. The similar issue can be seen here: https://github.com/mockito/mockito/issues/701

This is such a great pity to have to use much older and inferior version of Mockito... Can anything possibly be done to make it work with 2.x?

**Comment by NeverwinterMoon** [comment]
A small update. I did make it work with the old Mockito 1.10.19 version, but even then trying to stub mocked Realm's #close() still calls the real method right when I try to stub it:

```
val realmMock = Mockito.mock(Realm::class.java)
Mockito
  .doNothing()
  .`when`(realmMock)
  .close()
```

The above crashes on the spot as it tries to call the real method and throws `throw new IllegalStateException(INCORRECT_THREAD_CLOSE_MESSAGE)` as the Realm was never really opened.

**Comment by kneth** [comment]
@NeverwinterMoon Great to hear that you got it to work with an older version.

I am by no means a Mockito expert but can't you use partial mocking so `close()` isn't really called?

**Comment by NeverwinterMoon** [comment]
@kneth Sorry for the delay with my response, but that is exactly what I was trying to do.

From office Mockito docs on partial mocking (https://static.javadoc.io/org.mockito/mockito-core/2.4.2/org/mockito/Mockito.html#16):
```
//you can enable partial mock capabilities selectively on mocks:
Foo mock = mock(Foo.class);
//Be sure the real implementation is 'safe'.
//If real implementation throws exceptions or depends on specific state of the object then you're in trouble.
when(mock.someMethod()).thenCallRealMethod();
```

That is pretty much what I was doing - I wanted to `doNothing` if realm.close() was called.

Usually, if the class is public but the method that is being mock is not public, Mockito would not throw any exceptions but will call that method immediately when it trying to mock it. I am having a similar behavior, although I see that Realm#close is public...

**Comment by kneth** [comment]
It could be interesting to get some insights from the Mockito community on this.

**Comment by The0ldM0nk** [comment]
Guys, do we have any updates on this?

@NeverwinterMoon how did you finally fixed it? I am looking to fix a similar situation albeit a more complex one (rxjava slides in) and I am refraining from using `powermock` for now.

**Comment by NeverwinterMoon** [comment]
@The0ldM0nk 

I did try to go with PowerMock but soon gave up - the whole idea of PowerMock is not very appealing to me. The Mockito way wasn't working for me completely, too, when trying to run tests with Robolectric.

So, I ended up going with the Instrumented tests for pretty much everything. There, I used Mockito to mock only my own modules (all were injected with the Dagger in production code, so it was easy to mock them in tests) but used the actual Realm database.

AVD emulator got very fast lately, so running the instrumented tests on the emulators locally (or even Firebase test lab) is pretty fast. Also, it allows to run those tests on even the latest APIs. Robolectric was still limited to API level 21, the last time I checked. There are plenty of changes in the newer APIs...

**Comment by Zhuinden** [comment]
PowerMock should work with Robolectric 3.3.1+.
